### ISSUE TITLE
Provide a Default Policy Experience for Resiliency

### ISSUE THREAD

**Issue by halspang** [dapr/dapr#4512]
## In what area(s)?

/area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release

## Describe the proposal
With the initial launch of Resiliency in 1.7, we should provide an easier way for people to adopt it. Specifically, there will be a time when we want people to move to Resiliency and _remove_ the internal, built-in Dapr retries. This proposal is to get users to a place where they can more confidently adopt Resiliency without worrying that they'll lose all the retries that Dapr provides.

### Existing Retries
Dapr performs it's own retries in a few places already (this list may not be complete):

- Service invocation (specifically calling remote hosts)
- Actor invocation (specifically calling remote hosts)
- Actor state management (retries around storing reminders)
- Internal initialization (listing subscriptions, listing components)

Aside from the internal calls, the above calls can be covered by a Resiliency policy. As of 1.7, they exist in tandem with the Resiliency policy, which can cause more retries than intended.

The goal here will be to allow users to easily cover the existing retries and give them an avenue for using only Resiliency for all of their retry/backoff needs.

### Providing a Default Configuration
The easiest way to do this will be to provide a default configuration for Dapr's own retries. This configuration will use special names that we can recognize where retries exist and utilize them in our code base. This is necessary because Resiliency targets specific entities. The built-in retries are global. So, we'll need a way to make the policies recognize the global aspect of it.

This configuration can either be provided by default or setup by the customer. For an initial revision, it would be good to bundle it with Dapr so users are guaranteed their retries. Resiliency allows the setting of multiple policies (in k8s) and thus anything the customer wishes to set can be added without having to touch this configuration.

The policy itself will be provided in our documentation, but in Dapr, it will exist as part of the base level structure and be loaded into memory as a constant. The reason for this is that there has been a concern with users deleting/removing built-in retries entirely without intending to. By having it as an in-memory constant as part of the binary, the customer won't be able to delete it without explicitly overriding it. That can be accomplished by taking the documented configuration and changing the values which will override it, but not delete it. If a user goes through this much effort to override the built-ins we can assume they are a power user and responsible for their own actions.

### Utilizing the Configuration/Opting-Out of Built-ins
 Built-in retries don't always take place in the same area of code that the Resiliency policy retries do. So, in order to allow customers to utilize their own policy/opt-out of built-in retries we need to reference the policies in multiple places.

The new policies that we're defining are intended to be global as we don't know the name of every customer service/actor/component. As such, they can be seen as a default or fall-through policy. So, in the areas where we utilize them, we simply need to check to see if there is a user set policy covering the given target.

In short, we will utilize user defined policies first and only default to the built-in.

Yes, this does mean they can define a policy with less retries than the built-in, but they have actively defined the policy. A valid use case is for them to not want to retry certain things and that has to be supported.

## Summary
This proposal intends to:

- Provide an in-memory constant policy that represents the existing built-in retries
- Reference that policy in areas where we are performing built-in retries
- Let users override the defaults by setting their own
- Not use the default if a user has defined a more specific policy

**Comment by berndverst** [comment]
@halspang for service invocation -- could we also have circuitbreakers that can use a scope of per app, per path, or both? similar to how we have scopes for actors?

**Comment by halspang** [comment]
@berndverst mentioned that this sounds like it could cover all defaults, like all bindings/statestores/etc. The goal here is just to cover the built-in retries that Dapr already has, but as a future thought, we could use a similar thing here to cover all operations. This does let users simplify their policies if they have a lot of components for example.

Food for thought, not planning on targeting it here but I see the value in it for the future.

**Comment by halspang** [comment]
> @halspang for service invocation -- could we also have circuitbreakers that can use a scope of per app, per path, or both? similar to how we have scopes for actors?

That's something we could add. I think it'd be worth compiling an issue with ideas for future thoughts of Resiliency, in addition to the Phase 2 plans already in the original issue. I'll open one for discussion later :)

**Comment by yaron2** [comment]
> Aside from the internal calls, the above calls can be covered by a Resiliency policy. As of 1.7, they exist in tandem with the Resiliency policy, which can cause more retries than intended.

The service invocation API has an internal retry that needs to retry based on a very specific gRPC status code (Unauthenticated). How do we enable that with a Resiliency policy today?

**Comment by yaron2** [comment]
> The goal here will be to allow users to easily cover the existing retries and give them an avenue for using only Resiliency for all of their retry/backoff needs.

I'm not sure this sentence is accurate (specifically the `give them an avenue for using only Resiliency` part) because users do not proactively *use* any internal retries today. It's completely transparent to them and used to ensure the very basic reliability and/or stability of a call.

**Comment by halspang** [comment]
> The service invocation API has an internal retry that needs to retry based on a very specific gRPC status code (Unauthenticated). How do we enable that with a Resiliency policy today?

Resiliency policies just retry on errors, they don't care what they are except for 2 special cases for Circuit Breakers that indicate a permanent failure. What you can do though is:

```go
policy(func(ctx context.Context) error {
    err := grpcStuff()
    if errors.Is(err, grpc.Unauthenticated) {
        return err
    }
    return nil
})
```

> I'm not sure this sentence is accurate (specifically the give them an avenue for using only Resiliency part) because users do not proactively use any internal retries today. It's completely transparent to them and used to ensure the very basic reliability and/or stability of a call.

What I'm going for here is to bring the built-in retries into the resiliency space so that we can master everything under one path. Since we provide defaults, most people will never have to _use_ the built-in retries, but if you are so inclined you could now interact with it more directly through your resiliency policy.

**Comment by halspang** [comment]
To make that overwriting part a bit clearer, here's what I'm thinking terms of the default policy and how a customer can overwrite it. Yaron, if you don't think overwriting is a good idea for now, we can also reserve these names and error out if they try and overwrite them.

### In-Memory Constant
```yaml
apiVersion: dapr.io/v1alpha1
kind: Resiliency
metadata:
  name: resiliency
spec:
  policies:
    retries:
      DaprServiceInvocationRetry:
        policy: constant
        duration: 5s
        maxRetries: 10

      # These might be folded into 1 if their values in the runtime are the same.
      DaprActorInvocationRetry:
        policy: constant
        duration: 5s
        maxRetries: 10

targets:
    # Note that both targets only have retries.
    apps:
      DaprDefaulServiceTarget:
        retry: DaprServiceInvocationRetry

    actors:
      DaprDefaultActorTarget:
        retry: DaprActorInvocationRetry
```

### User Overwrite
```yaml
apiVersion: dapr.io/v1alpha1
kind: Resiliency
metadata:
  name: resiliency
spec:
  policies:
    timeouts:
      fiveSeconds: 5s

    retries:
      DaprServiceInvocationRetry:
        policy: constant
        duration: 5s
        maxRetries: 5 # Default now has 5 retries.

      retryForever:
        policy: exponential
        maxInterval: 15s
        maxRetries: 0 # Retry indefinitely

targets:
    # Note that both targets only have retries.
    apps:
      DaprDefaulServiceTarget:
        retry: DaprServiceInvocationRetry
        timeout: fiveSeconds # Introduce a timeout to default service invocation.

    actors:
      DaprDefaultActorTarget:
        retry: retryForever # Change the retry policy for default actors.
```

So like I said, the goal is to allow users full control, but to do everything for them if they don't want to actually engage. Thoughts, @yaron2? Again, if you don't think it's wise for us to be able to overwrite the defaults, we can reserve these keywords. Though I still think we shouldn't use the policy at all if the user has defined one for that specific target.

@berndverst - This is an example of what I was talking about ^

**Comment by halspang** [comment]
@artursouza / @yaron2 - Any more thoughts here before I get started on an implementation?

**Comment by mukundansundar** [comment]
>In short, we will utilize user defined policies first and only default to the built-in.

>Yes, this does mean they can define a policy with less retries than the built-in, but they have actively defined the policy. A valid use case is for them to not want to retry certain things and that has to be supported.

This does mean that they have to define the policy for that building block specifically otherwise the default policy would be used right? As in if user defines policy only for Actors, the default policy for service invocation should not be ignored. Is that correct? 

> Not use the default if a user has defined a more specific policy

In 1.7 if my understanding is correct, even if a more specific policy is defined, the default retries exist for the more generic use case. For example even if a user defines service invocation policy for specific remote hosts, the default retry policy of constant retries still exists and is executed for other remote hosts. Isn't that the case today?

**Comment by halspang** [comment]
@mukundansundar 

> This does mean that they have to define the policy for that building block specifically otherwise the default policy would be used right? As in if user defines policy only for Actors, the default policy for service invocation should not be ignored. Is that correct?

That is correct. We'll use the default policy where they don't define a more specific one. So if they only specify actors we'll use the default for service. It is worth noting though, that this is trying to cover just the retries that exist, it doesn't intend to give a full "backup" policy for every component. That being said though, it seems a logical next step after this as this code would enable it.

> In 1.7 if my understanding is correct, even if a more specific policy is defined, the default retries exist for the more generic use case. For example even if a user defines service invocation policy for specific remote hosts, the default retry policy of constant retries still exists and is executed for other remote hosts. Isn't that the case today?

Also correct. We decided that without a good way to migrate away from built-ins we would rather do more retries as opposed to accidentally no retries.
